### DISCUSSION TITLE
How does Slint handle Colors, e.g. for E-Ink displays

### DISCUSSION THREAD

**Discussion by pbert519** [slint-ui/slint#3421] [General]
Is it possible to define the dataformat / range for colors?
If we want to use a E-Ink display, that only supports 16 grayscales, i don't want to specify all kinds of colors within the ui and have to map them to different gray scales in the platform backend.
This has also a import on memory footprint, 16 grayscales fit in a u8, or even 4 bits, RGB colors usally use much more memory for each pixel.

**Comment by ogoffart** [comment]
In that case, you'll use the Slint's software renderer.
It has a `TargetPixel` trait that can be used to customize the format of the color. https://docs.rs/slint/latest/slint/platform/software_renderer/trait.TargetPixel.html
But that type must be at least one byte large.

What I suggest doing is to render into a 8bit grayscale pixel, and then convert to 4bit as you send the pixel to the screen.

For example: this is the backend for a monochrome display: 
https://github.com/ogoffart/carousel-mxchip-iot-devkit/blob/fb8c3948605cdbb77f7cdd7553ee1818c56de62e/src/main.rs#L125
It first render into a `sturct GrayPixel(u8)` that implements TargetPixel,  then convert to monochrome using a threshold.

**Reply by pbert519** [reply]
Thanks for the fast response!
I checked the example provided by you.

Using the TargetPixel with a u8, I'll continue to define full RGB colors within the slint files, which will then converted at runtime by the conversion functions in the trait implementation.
Or is this conversion part of compile process of the slint files?

And it's up to the UI developer to use only the RGB color values which directly match the grayscale values, e.g. by provided constants for each grayscale color?
Or expect the colors to be converted to next grayscale, as defined by the conversion functions.
